lib/utils/ec2.go

```diff
@@ -84,3 +88,71 @@ func IsEC2NodeID(id string) bool {
 func NodeIDFromIID(iid *imds.InstanceIdentityDocument) string {
 	return iid.AccountID + "-" + iid.InstanceID
 }
+
+// InstanceMetadataClient is a wrapper for an imds.Client.
+type InstanceMetadataClient struct {
+	c *imds.Client
+}
+
+// NewInstanceMetadataClient creates a new instance metadata client.
+func NewInstanceMetadataClient(ctx context.Context) (*InstanceMetadataClient, error) {
+	cfg, err := config.LoadDefaultConfig(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &InstanceMetadataClient{
+		c: imds.NewFromConfig(cfg),
+	}, nil
+}
+
+// IsAvailable checks if instance metadata is available.
+func (client *InstanceMetadataClient) IsAvailable(ctx context.Context) bool {
+	_, err := client.getMetadata(ctx, "")
+	return err == nil
+}
+
+// getMetadata gets the raw metadata from a specified path.
+func (client *InstanceMetadataClient) getMetadata(ctx context.Context, path string) (string, error) {
+	output, err := client.c.GetMetadata(ctx, &imds.GetMetadataInput{Path: path})
+	if err != nil {
+		return "", trace.Wrap(parseMetadataClientError(err))
+	}
+	defer output.Content.Close()
+	body, err := io.ReadAll(output.Content)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	return string(body), nil
+}
+
+// GetTagKeys gets all of the EC2 tag keys.
+func (client *InstanceMetadataClient) GetTagKeys(ctx context.Context) ([]string, error) {
+	body, err := client.getMetadata(ctx, "tags/instance")
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return strings.Split(body, "\n"), nil
+}
+
+// GetTagValue gets the value for a specified tag key.
+func (client *InstanceMetadataClient) GetTagValue(ctx context.Context, key string) (string, error) {
+	body, err := client.getMetadata(ctx, fmt.Sprintf("tags/instance/%s", key))
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	return body, nil
+}
+
+// parseMetadataClientError attempts to convert a failed imds call to a trace error.
+// If it can't, it returns the original error instead.
+func parseMetadataClientError(err error) error {
+	opError, ok := err.(*smithy.OperationError)
+	if !ok {
+		return err
+	}
+	httpErr, ok := opError.Err.(*smithyhttp.ResponseError)
+	if !ok {
+		return err
+	}
+	return trace.ReadError(httpErr.HTTPStatusCode(), nil)
```

Comment:
You should be able to simplify this by unwrapping the error without explicitly relying on smithy
```suggestion
	var httpError interface{ HTTPStatusCode() int }
	if errors.As(err, &httpError) {
		return trace.ReadError(httpErr.HTTPStatusCode(), nil)
	}
	return trace.Wrap(err)
```